src/custom-gutter-component.coffee

```diff
@@ -0,0 +1,109 @@
+{setDimensionsAndBackground} = require './gutter-component-helpers'
+
+# This class represents a gutter other than the 'line-numbers' gutter.
+# The contents of this gutter may be specified by Decorations.
+
+module.exports =
+class CustomGutterComponent
+
+  constructor: ({@gutter}) ->
+    @decorationNodesById = {}
+    @decorationItemsById = {}
+    @visible = true
+
+    @domNode = atom.views.getView(@gutter)
+    @decorationsNode = @domNode.firstChild
+
+  getDomNode: ->
+    @domNode
+
+  getName: ->
+    @gutter.name
+
+  hideNode: ->
+    if @visible
+      @domNode.style.display = 'none'
+      @visible = false
+
+  showNode: ->
+    if !@visible
+      @domNode.style.removeProperty('display')
+      @visible = true
+
+  updateSync: (state) ->
+    @oldDimensionsAndBackgroundState ?= {}
+    newDimensionsAndBackgroundState = state.gutters
+    setDimensionsAndBackground(@oldDimensionsAndBackgroundState, newDimensionsAndBackgroundState, @decorationsNode)
+
+    decorationState = state.gutters.customDecorations[@getName()]
+    return if !decorationState
+
+    updatedDecorationIds = new Set
+    for decorationId, decorationInfo of decorationState
+      updatedDecorationIds.add(decorationId)
+      existingDecoration = @decorationNodesById[decorationId]
+      if existingDecoration
+        @updateDecorationHTML(existingDecoration, decorationId, decorationInfo)
+      else
+        newNode = @buildDecorationHTML(decorationId, decorationInfo)
+        @decorationNodesById[decorationId] = newNode
+        @decorationsNode.appendChild(newNode)
+
+    for decorationId, decorationNode of @decorationNodesById
+      if !updatedDecorationIds.has(decorationId)
+        decorationNode.remove()
+        delete @decorationNodesById[decorationId]
+        delete @decorationItemsById[decorationId]
+
+  ###
+  Section: Private Methods
+  ###
+
+  # Builds and returns an HTMLElement to represent the specified decoration.
+  buildDecorationHTML: (decorationId, decorationInfo) ->
+    newNode = document.createElement('div')
+    newNode.classList.add('decoration')
+    newNode.style.top = decorationInfo.top + 'px'
+    newNode.style.height = decorationInfo.height + 'px'
+    newNode.style.position = 'absolute'
+    if decorationInfo.class
+      newNode.classList.add(decorationInfo.class)
+    @setDecorationItem(decorationInfo.item, decorationInfo.height, decorationId, newNode)
+    newNode
+
+  # Updates the existing HTMLNode with the new decoration info. Attempts to
+  # minimize changes to the DOM.
+  updateDecorationHTML: (existingNode, decorationId, newDecorationInfo) ->
+    if existingNode.style.top isnt newDecorationInfo.top + 'px'
```

Comment:
Reading these properties from the DOM is a no-no, because DOM reads can cause reflows. What you need to do instead is keep an `oldState` cache and compare against that instead.